api/src/wfl/module/covid.clj

```diff
@@ -23,6 +24,61 @@
 (defn create-covid-workload!
   [tx {:keys [source sink executor] :as request}])
 
+(defn start-covid-workload!
+  [{:keys [submission-id workspace] :as workload}])
+
+(defn ^:private snapshot-error-map
+  "Return default ExceptionInfo map for snapshot-related exceptions."
+  [{:keys [executor items] :as _workload} {:keys [id snapshot_id] :as _record}]
+  {:workspace (:workspace executor)
+   :table items
+   :id id
+   :snapshot_id snapshot_id})
+
+(defn ^:private try-get-snapshot-reference
+  "Attempt to fetch snapshot reference with id SNAPSHOT_REFERENCE_ID
+  from EXECUTOR-derived workspace."
+  [{:keys [executor] :as workload} {:keys [snapshot_reference_id] :as record}]
+  (let [workspace (:workspace executor)]
+    (try
+      (rawls/get-snapshot-reference workspace snapshot_reference_id)
+      (catch Throwable cause
+        (throw (ex-info "Rawls unable to fetch snapshot reference"
+                        (util/deep-merge
+                         (snapshot-error-map workload record)
+                         {:snapshot_reference_id snapshot_reference_id})
+                        cause))))))
+
+(defn ^:private snapshot-imported?
+  "Check if snapshot with id SNAPSHOT_ID in TDR has been imported to
+  workspace and can be successfully fetched."
+  [workload {:keys [snapshot_id snapshot_reference_id] :as record}]
+  {:pre [(some? snapshot_id)]}
+  (and (some? snapshot_reference_id)
+       (some? (try-get-snapshot-reference workload record))))
+
+(defn ^:private import-snapshot!
+  "Import snapshot with id SNAPSHOT_ID from TDR to EXECUTOR-derived workspace.
+  Use transaction TX to update ITEMS table with resulting reference id."
+  [tx {:keys [executor items] :as workload} {:keys [id snapshot_id] :as record}]
+  (let [workspace (:workspace executor)
+        name (util/randomize "placeholder-snapshot-ref-name")
```

Comment:
`"placeholder-snapshot-ref-name"` could blame `"wfl"`
to make the source clearer to anyone who finds it later.

Reply:
That's an interesting thought.  I've since updated the code to generate the snapshot reference name from the TDR snapshot name in `rawls/create-snapshot-reference` but we could supplement that naming to tie it back to `"wfl"` without much trouble.